lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -6,41 +6,134 @@
 #include "llvm/Support/raw_ostream.h"
 
 using namespace expr;
-
 namespace {
 // local code in this namespace
-
 //==-- main implementation goes here --==//
 class TCVR {
 
+  ExprMap m_cache;
+  std::map<Expr, ExprSet> m_endExpr;
+
+  bool m_isWellFormed;
+  Expr m_errorExp;
+
+  TypeChecker *const m_tc;
+  Expr m_topMost;
+
+  void reset(Expr exp) {
+    if (m_isWellFormed)
+      m_errorExp = Expr();
+
+    m_topMost = Expr();
+    m_isWellFormed = true;
+    m_cache.clear();
+  }
+
+  bool isConst(Expr exp) {
+    return bind::isBoolConst(exp) || bind::isIntConst(exp) ||
+           bind::isRealConst(exp) || bind::isUnintConst(exp) ||
+           bind::isArrayConst(exp) || bind::isStructConst(exp) ||
+           bv::isBvConst(exp);
+  }
+
+  bool isValue(Expr exp) { return isOpX<TRUE>(exp) || isOpX<FALSE>(exp); }
+
+  bool isSimpleType(Expr exp) {
+    return isConst(exp) || isValue(exp) || bv::is_bvnum(exp);
+  }
+
+  void mapEndTypes(Expr exp) {
```

Comment:
I don't understand this function. Creating a map from expressions to sets sounds quite expensive. I need to know more about what it is for.